plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemoryUsage.hs

```diff
@@ -0,0 +1,235 @@
+   things whose sizes are multiples of 8. -}
+instance ExMemoryUsage BS.ByteString where
+  -- Don't use `div` here!  That gives 1 instead of 0 for n=0.
+  memoryUsage bs = CostRose (unsafeToSatInt $ ((n - 1) `quot` 8) + 1) [] where
+      n = BS.length bs
+  {-# INLINE memoryUsage #-}
+
+instance ExMemoryUsage T.Text where
+  -- This is slow and inaccurate, but matches the version that was originally deployed.
+  -- We may try and improve this in future so long as the new version matches this exactly.
+  memoryUsage text = memoryUsage $ T.unpack text
+  {-# INLINE memoryUsage #-}
+
+instance ExMemoryUsage Int where
+  memoryUsage _ = CostRose 1 []
+  {-# INLINE memoryUsage #-}
+
+instance ExMemoryUsage Char where
+  memoryUsage _ = CostRose 1 []
+  {-# INLINE memoryUsage #-}
+
+instance ExMemoryUsage Bool where
+  memoryUsage _ = CostRose 1 []
+  {-# INLINE memoryUsage #-}
+
+instance ExMemoryUsage a => ExMemoryUsage [a] where
+    memoryUsage = CostRose 0 . map memoryUsage
+    {-# INLINE memoryUsage #-}
+
+{- Another naive traversal for size.  This accounts for the number of nodes in
+   a Data object, and also the sizes of the contents of the nodes.  This is not
+   ideal, but it seems to be the best we can do.  At present this only comes
+   into play for 'equalsData', which is implemented using the derived
+   implementation of '==' (fortunately the costing functions are lazy, so this
+   won't be called for things like 'unBData' which have constant costing
+   functions because they only have to look at the top node).  The problem is
+   that when we call 'equalsData' the comparison will take place entirely in
+   Haskell, so the costing functions for the contents of 'I' and 'B' nodes
+   won't be called.  Thus if we just counted the number of nodes the sizes of
+   'I 2' and 'B <huge bytestring>' would be the same but they'd take different
+   amounts of time to compare.  It's not clear how to trade off the costs of
+   processing a node and processing the contents of nodes: the implementation
+   below compromises by charging four units per node, but we may wish to revise
+   this after experimentation.
+-}
+instance ExMemoryUsage Data where
+    memoryUsage = sizeData where
+        -- The cost of each node of the 'Data' object (in addition to the cost of its content).
+        nodeMem = CostRose 4 []
+        {-# INLINE nodeMem #-}
+
+        -- Add two 'CostRose's. We don't make this into a 'Semigroup' instance, because there exist
+        -- different ways to add two 'CostRose's (e.g. we could optimize the case when one of the
+        -- roses contains only one element or we can make the function lazy in the second argument).
+        -- Here we chose the version that is most efficient when the first argument is @nodeMem@ (we
+        -- didn't do any benchmarking though, so it may not be the most efficient one) -- we don't
+        -- have any other cases.
+        combine (CostRose cost1 forest1) (CostRose cost2 forest2) =
```

Comment:
It doesn't, but I felt like it would less clear if it wasn't this general and `nodeMem` was hardcoded in there. It does indeed get inlined and optimized properly.